### DISCUSSION TITLE
Manage MultiCluster in a single repo

### DISCUSSION THREAD

**Discussion by jcamu** [fluxcd/flux2#2437] [Q&A] [answered]
Hello,

We are currently evaluating different solution to manage all our new Kubernetes Infrastructures in a GitOps way.
ArgoCD, Flux and Fleet seems to be the trending solutions.

We still have a clear separation between ops and dev, we are only looking for a solution for infra code , devs will continue (for now) to use dedicated pipelines.

Reading through the different documentation we are not sure if flux2 should be used for this, it is maybe not the aim of this solution.
Most of the time, the documentation, demo show the solution used for one team and several applications in few env (dev, test, prod).
But in our case, we want to have one repository containing all our clusters (10-100) configuration, for lots of different teams with dedicated clusters. 
What is the limit to the number of clusters linked to one repo? 

In addition, in the best practice, the Flux repo should be in a separate repo than the IAC cluster provisionning solution like Terraform or in the same?
The pipeline creating the cluster should also initiate the flux operator in the new cluster i suppose.

Maybe this question was already raised and answered, but are not sure of the use case.
Thanks.

Regards,

**Comment by stefanprodan** [comment, marked as answer]
> What is the limit to the number of clusters linked to one repo?

There is no limit on the Flux side.

> In addition, in the best practice, the Flux repo should be in a separate repo than the IAC cluster provisionning solution like Terraform or in the same?

It's up to you, in the https://github.com/fluxcd/flux2-multi-tenancy example, the Flux manifests for all clusters along with the cluster addons are in one repo, while the "tenants" are in their dedicated repos. Tenants can mean anything from teams to apps to microservices.

> The pipeline creating the cluster should also initiate the flux operator in the new cluster i suppose.

You can use the [Flux CLI](https://fluxcd.io/docs/installation/#bootstrap) in your cluster creation pipeline or use the [Flux Terraform provider](https://github.com/fluxcd/terraform-provider-flux) or use the [Flux Helm chart](https://artifacthub.io/packages/helm/fluxcd-community/flux2).

**Reply by jcamu** [reply]
Thanks for your quick reply

**Reply by stealthybox** [reply]
Flux works very well for this use case.
Each cluster can have its own credential for reading from a repo.
Each cluster can be configured with a different folder and/or branch/tag/commit-ref, so that a single repo can describe many different types of clusters through the use of patches or distinct manifests.

**Reply by makkes** [reply]
In addition to what @stefanprodan and @stealthybox already said you can have a look at the PoC that I conducted and maintain at https://github.com/makkes/flux-mc-control-plane for a more complex setup. It involves a central repository for managing the clusters but uses different "sub-repos" if you will that individual teams can be given access to because you might not want to give all teams access to a single repository for several reasons.
